File: C_Compiler_Scratch/parse.py
import sys
from lex import *

class Parser:
    '''
    The Parser is primarily responsible for checking if the grammar of the source code is valid according to the grammar rules we define in the parser. It also outputs the C code with the help of the emitter. The parser uses Recursive Descent. It receives one token from the lexer, at a time.
    '''
    def __init__(self, lexer, emitter):
        self.lexer = lexer
        self.emitter = emitter

        self.symbols = set()    #Keep track of variables
        self.labelsDeclared = set() # Keep track of all labels declared
        self.labelsGotoed = set() # All labels goto'ed so far

        self.curToken = None
        self.peekToken = None
        self.nextToken()
        self.nextToken()  #Next token written such that peekToken passes the current token to curToken, then peakToken inherits the next token passed to it by the lexer.

    #Return true if the current token matches the token required by the grammar rule
    def checkToken(self, kind):
        return kind == self.curToken.kind

    # Return true if the next token matches.
    def checkPeek(self, kind):
        return kind == self.peekToken.kind

    #Crucial function to ensure that specific tokens come after other tokens in accordance to the grammar rule. Error if not.
    def match(self, kind):
        if not self.checkToken(kind):
            self.abort("Expected " + kind.name + ", got " + self.curToken.kind.name)
        self.nextToken()

    def nextToken(self):
        self.curToken = self.peekToken
        self.peekToken = self.lexer.getToken()       

    # Return true if the current token is a comparison operator.    
    def isComparisonOperator(self): 
        COMP_OPERATORS = {TokenType.GT,
                    TokenType.GTEQ,
                    TokenType.LT,
                    TokenType.LTEQ,
                    TokenType.EQEQ,
                    TokenType.NOTEQ}       
        return self.curToken.kind in COMP_OPERATORS

    def abort(self, message):
        sys.exit("Error! " + message)


    



    # program ::= {statement}
    def program(self):
        '''The program function is responsible for running the parser'''
        self.emitter.headerLine("#include <stdio.h>") #Every emitted C code will begin with this code, hence its at the top of the program function
        self.emitter.headerLine("int main(void){")
        
        # Skip Newlines
        while self.checkToken(TokenType.NEWLINE):
            self.nextToken()

        # Parse all the statements in the program.
        while not self.checkToken(TokenType.EOF):
            self.statement() #Runs statement function, which in turn Parses all the incoming code

        # Every emitted C code will end with this code.
        self.emitter.emitLine("return 0;")
        self.emitter.emitLine("}")

        # Check that each label referenced in a GOTO is declared.
        for label in self.labelsGotoed:
            if label not in self.labelsDeclared:
                self.abort("Attempting to GOTO to undeclared label: " + label)


    
    def statement(self):
        '''The statement function is the one of the enforcers of the grammar rules of the source code. We are ensuring that tokens are coming in valid sequences eg IF is followed by THEN.'''
    
        # IF Current Token is PRINT
        if self.checkToken(TokenType.PRINT):
            self.nextToken()

            
            if self.checkToken(TokenType.STRING):
                #Convert to C code
                self.emitter.emitLine("printf(\"" + self.curToken.text + "\\n\");")
                self.nextToken()

            else:
                # Expect an expression and print the result as a float
                self.emitter.emit("printf(\"%" + ".2f\\n\", (float)(")
                self.expression()
                self.emitter.emitLine("));")

        # If Current Token is IF
        elif self.checkToken(TokenType.IF):
            self.nextToken()
            self.emitter.emit("if(")
            self.comparison() #Check if whatever comes after IF passes the grammar check using comparison

            self.match(TokenType.THEN) #Match Ensures that IF ends with THEN
            self.nl()
            self.emitter.emitLine("){")

            # Recursive call to run statement on whatever is inside the body of the IF statement
            while not self.checkToken(TokenType.ENDIF):
                self.statement()

            self.match(TokenType.ENDIF)
            self.emitter.emitLine("}")

        # If Current Token is WHILE
        elif self.checkToken(TokenType.WHILE):
            self.nextToken() #Go to next token to check if whatever comes after while adheres to grammar rules.
            self.emitter.emit("while(") #Emit the C code equivalent
            self.comparison()

            self.match(TokenType.REPEAT)
            self.nl()
            self.emitter.emitLine("){")

            # # Recursive call to run statement on whatever is inside the body of the WHILE statement
            while not self.checkToken(TokenType.ENDWHILE):
                self.statement()

            self.match(TokenType.ENDWHILE)
            self.emitter.emitLine("}")

        # If Current Token is LABEL
        elif self.checkToken(TokenType.LABEL):
            self.nextToken()

            # Make sure this label doesn't already exist.
            if self.curToken.text in self.labelsDeclared:
                self.abort("Label already exists: " + self.curToken.text)
            self.labelsDeclared.add(self.curToken.text)

            self.emitter.emitLine(self.curToken.text + ":")
            self.match(TokenType.IDENT)

        # If Current Token is GOTO
        elif self.checkToken(TokenType.GOTO):
            self.nextToken()
            self.labelsGotoed.add(self.curToken.text)
            self.emitter.emitLine("goto " + self.curToken.text + ";")
            self.match(TokenType.IDENT)

        # If Current Token is LET
        elif self.checkToken(TokenType.LET):
            self.nextToken()

            #  Check if ident exists in symbol table. If not, declare it.
            if self.curToken.text not in self.symbols:
                self.symbols.add(self.curToken.text)
                self.emitter.headerLine("float " + self.curToken.text + ";")

            self.emitter.emit(self.curToken.text + " = ")
            self.match(TokenType.IDENT)
            self.match(TokenType.EQ)
            
            self.expression()
            self.emitter.emitLine(";")

        # If Current Token is INPUT
        elif self.checkToken(TokenType.INPUT):
            self.nextToken()

            # If variable doesn't already exist, declare it.
            if self.curToken.text not in self.symbols:
                self.symbols.add(self.curToken.text)
                self.emitter.headerLine("float " + self.curToken.text + ";")

            # Emit scanf but also validate the input. If invalid, set the variable to 0 and clear the input.
            self.emitter.emitLine("if(0 == scanf(\"%" + "f\", &" + self.curToken.text + ")) {")
            self.emitter.emitLine(self.curToken.text + " = 0;")
            self.emitter.emit("scanf(\"%")
            self.emitter.emitLine("*s\");")
            self.emitter.emitLine("}")
            self.match(TokenType.IDENT)

        else:
            self.abort("Invalid statement at " + self.curToken.text + " (" + self.curToken.kind.name + ")")

        self.nl()


    # comparison ::= expression (("==" | "!=" | ">" | ">=" | "<" | "<=") expression)+
    def comparison(self):
        self.expression()
        # Must be at least one comparison operator and another expression.
        if self.isComparisonOperator():
            self.emitter.emit(self.curToken.text)
            self.nextToken()
            self.expression()
        # Can have 0 or more comparison operator and expressions.
        while self.isComparisonOperator():
            self.emitter.emit(self.curToken.text)
            self.nextToken()
            self.expression()


    # expression ::= term {( "-" | "+" ) term}
    def expression(self):
        self.term()
        # Can have 0 or more +/- and expressions.
        while self.checkToken(TokenType.PLUS) or self.checkToken(TokenType.MINUS):
            self.emitter.emit(self.curToken.text)
            self.nextToken()
            self.term()


    # term ::= unary {( "/" | "*" ) unary}
    def term(self):
        self.unary()
        # Can have 0 or more *// and expressions.
        while self.checkToken(TokenType.ASTERISK) or self.checkToken(TokenType.SLASH):
            self.emitter.emit(self.curToken.text)
            self.nextToken()
            self.unary()


    # unary ::= ["+" | "-"] primary
    def unary(self):
        # Optional unary +/-
        if self.checkToken(TokenType.PLUS) or self.checkToken(TokenType.MINUS):
            self.emitter.emit(self.curToken.text)
            self.nextToken()        
        self.primary()


    # primary ::= number | ident
    def primary(self):
        if self.checkToken(TokenType.NUMBER): 
            self.emitter.emit(self.curToken.text)
            self.nextToken()
        elif self.checkToken(TokenType.IDENT):
            # Ensure the variable already exists.
            if self.curToken.text not in self.symbols:
                self.abort("Referencing variable before assignment: " + self.curToken.text)

            self.emitter.emit(self.curToken.text)
            self.nextToken()
        else:
            # Error!
            self.abort("Unexpected token at " + self.curToken.text)

    # nl ::= '\n'+
    def nl(self):
        
        # But we will allow extra newlines too, of course.
        while self.checkToken(TokenType.NEWLINE):
            self.nextToken()

File: C_Compiler_Scratch/parse.cpp
#include "parse.h"
#include <iostream>
#include <string>

Parser::Parser(Lexer& lex, Emitter& emit) 
    : lexer(lex), 
      emitter(emit), 
      currentToken("", TokenType::EOF_TOKEN), 
      peekToken("", TokenType::EOF_TOKEN) {
    nextToken();
    nextToken(); // Call twice to initialize currentToken and peekToken
}

void Parser::nextToken() {
    currentToken = peekToken;
    peekToken = lexer.getToken();
}

void Parser::match(TokenType kind) {
    if (currentToken.getKind() != kind) {
        abort("Expected " + std::to_string(static_cast<int>(kind)) + 
              ", got " + std::to_string(static_cast<int>(currentToken.getKind())));
    }
    nextToken();
}

void Parser::abort(const std::string& message) {
    std::cerr << "Error! " << message << std::endl;
    exit(1);
}

void Parser::program() {
    // Write the C program header
    emitter.emitLine("#include <stdio.h>");
    emitter.emitLine("int main(void) {");
    
    // Skip leading newlines if any
    while (currentToken.getKind() == TokenType::NEWLINE) {
        nextToken();
    }
    
    // Parse statements until the end of the file
    while (currentToken.getKind() != TokenType::EOF_TOKEN) {
        statement();
        
        // Skip any newlines between statements
        while (currentToken.getKind() == TokenType::NEWLINE) {
            nextToken();
        }
    }
    
    // Write the C program footer
    emitter.emitLine("    return 0;");
    emitter.emitLine("}");
}

void Parser::statement() {
    // Check the first token to determine statement type
    
    // "PRINT" (expression | string)
    if (currentToken.getKind() == TokenType::PRINT) {
        nextToken();
        
        if (currentToken.getKind() == TokenType::STRING) {
            // Simple string
            emitter.emitLine("    printf(\"" + currentToken.getText() + "\\n\");");
            nextToken();
        } else {
            // Expression
            emitter.emit("    printf(\"%d\\n\", ");
            expression();
            emitter.emitLine(");");
        }
    }
    // "IF" comparison "THEN" nl {statement} "ENDIF"
    else if (currentToken.getKind() == TokenType::IF) {
        nextToken();
        emitter.emit("    if(");
        comparison();
        
        match(TokenType::THEN);
        emitter.emitLine(") {");
        
        // Skip any newlines after THEN
        while (currentToken.getKind() == TokenType::NEWLINE) {
            nextToken();
        }
        
        // Zero or more statements in the body
        while (currentToken.getKind() != TokenType::ENDIF) {
            statement();
            
            // Skip any newlines between statements
            while (currentToken.getKind() == TokenType::NEWLINE) {
                nextToken();
            }
        }
        
        match(TokenType::ENDIF);
        emitter.emitLine("    }");
    }
    // "WHILE" comparison "REPEAT" nl {statement} "ENDWHILE"
    else if (currentToken.getKind() == TokenType::WHILE) {
        nextToken();
        emitter.emit("    while(");
        comparison();
        
        match(TokenType::REPEAT);
        emitter.emitLine(") {");
        
        // Skip any newlines after REPEAT
        while (currentToken.getKind() == TokenType::NEWLINE) {
            nextToken();
        }
        
        // Zero or more statements in the body
        while (currentToken.getKind() != TokenType::ENDWHILE) {
            statement();
            
            // Skip any newlines between statements
            while (currentToken.getKind() == TokenType::NEWLINE) {
                nextToken();
            }
        }
        
        match(TokenType::ENDWHILE);
        emitter.emitLine("    }");
    }
    // "LET" ident = expression
    else if (currentToken.getKind() == TokenType::LET) {
        nextToken();
        
        if (currentToken.getKind() != TokenType::IDENT) {
            abort("Expected identifier after 'LET'");
        }
        
        std::string varName = currentToken.getText();
        nextToken();
        
        match(TokenType::EQ);
        
        emitter.emit("    " + varName + " = ");
        expression();
        emitter.emitLine(";");
    }
    // "INPUT" ident
    else if (currentToken.getKind() == TokenType::INPUT) {
        nextToken();
        
        if (currentToken.getKind() != TokenType::IDENT) {
            abort("Expected identifier after 'INPUT'");
        }
        
        std::string varName = currentToken.getText();
        emitter.emitLine("    printf(\"Input: \");");
        emitter.emitLine("    scanf(\"%d\", &" + varName + ");");
        nextToken();
    }
    // This is not a valid statement
    else {
        abort("Invalid statement at: " + currentToken.getText());
    }
    
    // Newline - make this optional by checking if the current token is a newline
    if (currentToken.getKind() == TokenType::NEWLINE) {
        nextToken();
    }
    // If not a newline, we don't need to match it
}

void Parser::comparison() {
    expression();
    
    // Must be a comparison operator
    if (currentToken.getKind() == TokenType::EQEQ || currentToken.getKind() == TokenType::NOTEQ ||
        currentToken.getKind() == TokenType::LT || currentToken.getKind() == TokenType::LTEQ ||
        currentToken.getKind() == TokenType::GT || currentToken.getKind() == TokenType::GTEQ) {
        
        std::string op;
        switch (currentToken.getKind()) {
            case TokenType::EQEQ:  op = "=="; break;
            case TokenType::NOTEQ: op = "!="; break;
            case TokenType::LT:    op = "<"; break;
            case TokenType::LTEQ:  op = "<="; break;
            case TokenType::GT:    op = ">"; break;
            case TokenType::GTEQ:  op = ">="; break;
            default: break; // Never reached
        }
        
        nextToken();
        emitter.emit(op + " ");
        expression();
    } else {
        abort("Expected comparison operator");
    }
}

void Parser::expression() {
    term();
    
    // Can have 0 or more +/- operations
    while (currentToken.getKind() == TokenType::PLUS || currentToken.getKind() == TokenType::MINUS) {
        emitter.emit(currentToken.getText() + " ");
        nextToken();
        term();
    }
}

void Parser::term() {
    unary();
    
    // Can have 0 or more *// operations
    while (currentToken.getKind() == TokenType::ASTERISK || currentToken.getKind() == TokenType::SLASH) {
        emitter.emit(currentToken.getText() + " ");
        nextToken();
        unary();
    }
}

void Parser::unary() {
    // Optional unary +/-
    if (currentToken.getKind() == TokenType::PLUS || currentToken.getKind() == TokenType::MINUS) {
        emitter.emit(currentToken.getText());
        nextToken();
    }
    
    primary();
}

void Parser::primary() {
    if (currentToken.getKind() == TokenType::NUMBER) {
        emitter.emit(currentToken.getText());
        nextToken();
    } else if (currentToken.getKind() == TokenType::IDENT) {
        emitter.emit(currentToken.getText());
        nextToken();
    } else {
        abort("Unexpected token: " + currentToken.getText());
    }
}
